live/lib/auth_app/accounts.ex

```diff
@@ -235,119 +196,116 @@ defmodule AuthApp.Accounts do
   """
-  def deliver_user_confirmation_instructions(%User{} = user, confirmation_url_fun)
-      when is_function(confirmation_url_fun, 1) do
-    if user.confirmed_at do
-      {:error, :already_confirmed}
+  def get_user_by_magic_link_token(token) do
+    with {:ok, query} <- UserToken.verify_magic_link_token_query(token),
+         {user, _token} <- Repo.one(query) do
+      user
     else
-      {encoded_token, user_token} = UserToken.build_email_token(user, "confirm")
-      Repo.insert!(user_token)
-      UserNotifier.deliver_confirmation_instructions(user, confirmation_url_fun.(encoded_token))
+      _ -> nil
     end
   end
 
   @doc """
-  Confirms a user by the given token.
+  Gets the user with the given signed token and deletes the used token.
+
+  Because of the way the magic link log in works, there are three cases to consider:
+
+    1. The user has not confirmed their email yet, no password is set.
+       In that case, the user gets confirmed and signed in.
+    2. The user has already confirmed their email. They are signed in.
+    3. The user has not confirmed their email yet, but a password is set.
+
+  The third case cannot happen with the default implementation, but it may happen
+  when adapting the code to a different use case, e.g. allowing to register with a password.
 
-  If the token matches, the user account is marked as confirmed
-  and the token is deleted.
+  This situation poses a security problem: an attacker may register with an email they
+  don't actually control which can lead to a session fixation attack:
+  The attacker can wait for the actual user to register the app. When they log in by magic link,
+  the account would get confirmed and the attacker would gain access by knowing the password
+  they set initially. To prevent this, the default implementation raises an error when trying
+  to use a magic link with a password set and unconfirmed email.
   """
-  def confirm_user(token) do
-    with {:ok, query} <- UserToken.verify_email_token_query(token, "confirm"),
-         %User{} = user <- Repo.one(query),
-         {:ok, %{user: user}} <- Repo.transaction(confirm_user_multi(user)) do
-      {:ok, user}
-    else
-      _ -> :error
+  def magic_link_login(token) do
+    {:ok, query} = UserToken.verify_magic_link_token_query(token)
+
+    case Repo.one(query) do
+      # prevent session fixation attacks by disallowing magic link log in
+      # for unconfirmed users with a password set
+      {%User{confirmed_at: nil, hashed_password: hash}, _token} when not is_nil(hash) ->
+        raise """
+        Magic link log in is not allowed for unconfirmed users with a password set!
+
+        This cannot happen with the default implementation, which indicates that you
+        might have adapted the code to a different use case. Please make sure to read the phx.gen.auth
+        documentation (mix help phx.gen.auth) about the security implications when allowing passwords
+        for unconfirmed users.
```

Comment:
```suggestion
        magic link log in is not allowed for unconfirmed users with a password set!

        This cannot happen with the default implementation, which indicates that you
        might have adapted the code to a different use case. Please make sure to read the phx.gen.auth
        documentation (mix help phx.gen.auth) about the security implications when allowing passwords
        for unconfirmed users
```

Messages start in lowercase